File: crates/ferromode-r/tests/testthat/test-binding.R
test_that("emd decomposes a sine wave", {
  signal <- sin(seq(0, 2 * pi, length.out = 100))
  result <- emd(signal, list())
  expect_type(result, "list")
  expect_true(result$n_imfs >= 1)
  expect_equal(result$n_samples, 100L)
  expect_equal(result$algorithm, "emd")
  expect_true(result$elapsed_ms >= 0)
})

test_that("emd respects max_imfs", {
  signal <- sin(seq(0, 2 * pi, length.out = 100))
  result <- emd(signal, list(max_imfs = 1L))
  expect_equal(result$n_imfs, 1L)
})

test_that("eemd produces results", {
  signal <- sin(seq(0, 2 * pi, length.out = 100))
  result <- eemd(signal, list(num_ensembles = 5L, noise_std = 0.2, seed = 42L))
  expect_type(result, "list")
  expect_true(result$n_imfs >= 1)
  expect_equal(result$algorithm, "eemd")
})

test_that("ceemd produces results", {
  signal <- sin(seq(0, 2 * pi, length.out = 100))
  result <- ceemd(signal, list(num_ensembles = 5L, noise_std = 0.2, seed = 42L))
  expect_type(result, "list")
  expect_true(result$n_imfs >= 1)
  expect_equal(result$algorithm, "ceemd")
})

test_that("ceemdan produces results", {
  signal <- sin(seq(0, 2 * pi, length.out = 100))
  result <- ceemdan(signal, list(num_ensembles = 5L, noise_std = 0.2, seed = 42L))
  expect_type(result, "list")
  expect_true(result$n_imfs >= 1)
  expect_equal(result$algorithm, "ceemdan")
})

test_that("iceemdan produces results", {
  signal <- sin(seq(0, 2 * pi, length.out = 100))
  result <- iceemdan(signal, list(num_ensembles = 5L, noise_std = 0.2, seed = 42L))
  expect_type(result, "list")
  expect_true(result$n_imfs >= 1)
  expect_equal(result$algorithm, "iceemdan")
})

test_that("vmd decomposes a two-tone signal", {
  t <- seq(0, 1, length.out = 200)
  signal <- sin(2 * pi * 5 * t) + 0.5 * sin(2 * pi * 20 * t)
  result <- vmd(signal, list(n_modes = 2L, alpha = 2000.0, tol = 1e-7, max_iterations = 500L))
  expect_type(result, "list")
  expect_equal(result$n_imfs, 2L)
  expect_equal(result$algorithm, "vmd")
})

test_that("reconstruct recovers original signal", {
  signal <- sin(seq(0, 2 * pi, length.out = 100))
  result <- emd(signal, list())
  reconstructed <- reconstruct(result)
  expect_equal(length(reconstructed), length(signal))
  expect_true(max(abs(signal - reconstructed)) < 1e-6)
})

test_that("ferromode_r_version returns a string", {
  ver <- ferromode_r_version()
  expect_type(ver, "character")
  expect_true(nchar(ver) > 0)
})

test_that("emd honors boundary + full config (palindrome_cyclic)", {
  signal <- sin(seq(0, 4 * pi, length.out = 128))
  result <- emd(signal, list(max_imfs = 4L, boundary_condition = "palindrome_cyclic",
                             spline_type = "periodic", energy_threshold = 1e-7))
  expect_true(result$n_imfs >= 1)
  expect_equal(result$algorithm, "emd")
})

test_that("memd decomposes multivariate channels", {
  ch1 <- sin(seq(0, 4 * pi, length.out = 128))
  ch2 <- cos(seq(0, 4 * pi, length.out = 128))
  result <- memd(list(ch1, ch2), list(num_directions = 16L, max_imfs = 3L))
  expect_equal(result$algorithm, "memd")
  expect_true(result$n_imfs >= 1)
})

test_that("namemd decomposes multivariate channels", {
  ch1 <- sin(seq(0, 4 * pi, length.out = 128))
  ch2 <- cos(seq(0, 4 * pi, length.out = 128))
  result <- namemd(list(ch1, ch2), list(num_directions = 16L, max_imfs = 3L,
                                        n_noise_channels = 2L, noise_std = 0.1, seed = 42L))
  expect_equal(result$algorithm, "namemd")
  expect_true(result$n_imfs >= 1)
})

test_that("hilbert returns spectra", {
  signal <- sin(seq(0, 8 * pi, length.out = 128))
  result <- emd(signal, list(max_imfs = 4L))
  h <- hilbert(result$imfs, 100.0)
  expect_equal(length(h$instantaneous_amplitude), result$n_imfs)
  expect_true(length(h$marginal_spectrum) > 0)
})

test_that("streaming decomposes chunks", {
  handle <- streaming_new(list(max_imfs = 4L), 2048L, 3L)
  chunk <- sin(seq(0, 4 * pi, length.out = 256))
  out <- streaming_decompose_chunk(handle, chunk)
  expect_true(length(out$imfs) >= 1)
  expect_true(is.numeric(out$spectral_entropy))
  streaming_reset(handle)
})

test_that("differentiable forward + backward", {
  signal <- sin(seq(0, 4 * pi, length.out = 200)) + 0.3 * seq(0, 1, length.out = 200)
  fwd <- emd_forward(signal, list(max_imfs = 4L))
  expect_true(length(fwd$imfs) >= 1)
  expect_true(is.finite(fwd$reconstruction_error))
  grads <- lapply(seq_along(fwd$imfs), function(i) rep(1.0, 200))
  grad <- emd_backward(grads, signal)
  expect_equal(length(grad), 200L)
  expect_true(abs(grad[1] - 1.0) < 1e-12)
})

test_that("emd rejects non-finite values", {
  signal <- c(1.0, NaN, 3.0, 4.0, 5.0)
  expect_error(emd(signal, list()), "non-finite")
})

test_that("emd rejects too-short signal", {
  signal <- c(1.0, 2.0)
  expect_error(emd(signal, list()), "insufficient data")
})

File: crates/ferromode-py/tests/test_binding.py
"""Smoke tests for ferromode Python bindings.

Tests assert output shape, dtype, and that .reconstruct() returns
an array of the correct length for each algorithm.
"""

import numpy as np
import pytest


def _make_signal(n: int = 200) -> np.ndarray:
    t = np.linspace(0, 1, n, endpoint=False)
    return np.sin(2 * np.pi * 5 * t) + 0.5 * np.sin(2 * np.pi * 20 * t)


def _make_multivariate_signal(n_channels: int = 3, n_samples: int = 200) -> np.ndarray:
    t = np.linspace(0, 1, n_samples, endpoint=False)
    channels = []
    for i in range(n_channels):
        freq = 5.0 + i * 10.0
        channels.append(np.sin(2 * np.pi * freq * t))
    return np.array(channels, dtype=np.float64)


# ─── EMD ───────────────────────────────────────────────────────────────


def test_emd_default_config():
    import ferromode

    signal = _make_signal()
    result = ferromode.emd(signal)
    assert result is not None
    imfs = result.imfs()
    assert imfs.n_imfs() >= 1
    recon = result.reconstruct()
    assert len(recon) == len(signal)
    assert recon.dtype == np.float64


def test_emd_custom_config():
    import ferromode

    signal = _make_signal()
    config = ferromode.EmdConfig(max_imfs=3)
    result = ferromode.emd(signal, config)
    imfs = result.imfs()
    assert imfs.n_imfs() <= 3
    recon = result.reconstruct()
    assert len(recon) == len(signal)


def test_emd_error_on_empty():
    import ferromode

    with pytest.raises(ValueError):
        ferromode.emd(np.array([], dtype=np.float64))


def test_emd_error_on_nan():
    import ferromode

    with pytest.raises(ValueError):
        ferromode.emd(np.array([1.0, np.nan, 3.0], dtype=np.float64))


# ─── EEMD ──────────────────────────────────────────────────────────────


def test_eemd_default():
    import ferromode

    signal = _make_signal()
    config = ferromode.EnsembleConfig(num_ensembles=5, seed=42)
    result = ferromode.eemd(signal, config)
    imfs = result.imfs()
    assert imfs.n_imfs() >= 1
    recon = result.reconstruct()
    assert len(recon) == len(signal)


def test_eemd_with_emd_config():
    import ferromode

    signal = _make_signal()
    ens = ferromode.EnsembleConfig(num_ensembles=5, seed=42)
    emd_cfg = ferromode.EmdConfig(max_imfs=5)
    result = ferromode.eemd(signal, ens, emd_cfg)
    imfs = result.imfs()
    assert imfs.n_imfs() >= 1


# ─── CEEMD ─────────────────────────────────────────────────────────────


def test_ceemd_default():
    import ferromode

    signal = _make_signal()
    config = ferromode.EnsembleConfig(num_ensembles=5, seed=42)
    result = ferromode.ceemd(signal, config)
    imfs = result.imfs()
    assert imfs.n_imfs() >= 1
    recon = result.reconstruct()
    assert len(recon) == len(signal)


# ─── CEEMDAN ───────────────────────────────────────────────────────────


def test_ceemdan_default():
    import ferromode

    signal = _make_signal()
    config = ferromode.EnsembleConfig(num_ensembles=10, seed=42)
    result = ferromode.ceemdan(signal, config)
    imfs = result.imfs()
    assert imfs.n_imfs() >= 1
    recon = result.reconstruct()
    assert len(recon) == len(signal)


# ─── ICEEMDAN ──────────────────────────────────────────────────────────


def test_iceemdan_default():
    import ferromode

    signal = _make_signal()
    config = ferromode.EnsembleConfig(num_ensembles=10, seed=42)
    result = ferromode.iceemdan(signal, config)
    imfs = result.imfs()
    assert imfs.n_imfs() >= 1
    recon = result.reconstruct()
    assert len(recon) == len(signal)


# ─── MEMD ──────────────────────────────────────────────────────────────


def test_memd_default():
    import ferromode

    signal = _make_multivariate_signal(n_channels=3, n_samples=200)
    config = ferromode.MemdConfig(num_directions=16, max_sifting_iterations=20)
    result = ferromode.memd(signal, config)
    imfs = result.imfs()
    assert imfs.n_imfs() >= 1
    recon = result.reconstruct()
    # MEMD residue is flattened across channels
    assert len(recon) > 0


# ─── NA-MEMD ───────────────────────────────────────────────────────────


def test_namemd_default():
    import ferromode

    signal = _make_multivariate_signal(n_channels=3, n_samples=200)
    config = ferromode.NaMemdConfig(
        n_noise_channels=2,
        noise_std=0.1,
        seed=42,
        num_directions=16,
        max_sifting_iterations=20,
    )
    result = ferromode.namemd(signal, config)
    imfs = result.imfs()
    assert imfs.n_imfs() >= 1
    recon = result.reconstruct()
    assert len(recon) > 0


# ─── VMD ───────────────────────────────────────────────────────────────


def test_vmd_default():
    import ferromode

    signal = _make_signal()
    config = ferromode.VmdConfig(n_modes=2, max_iterations=100)
    result = ferromode.vmd(signal, config)
    imfs = result.imfs()
    assert imfs.n_imfs() == 2
    recon = result.reconstruct()
    assert len(recon) == len(signal)


def test_vmd_error_on_too_many_modes():
    import ferromode

    signal = _make_signal(n=20)
    config = ferromode.VmdConfig(n_modes=15)
    with pytest.raises(ValueError):
        ferromode.vmd(signal, config)


# ─── Hilbert transform ────────────────────────────────────────────────


def test_hilbert_transform():
    import ferromode

    signal = _make_signal()
    result = ferromode.emd(signal)
    hilbert = result.hilbert(sample_rate=1000.0)
    amp = hilbert.instantaneous_amplitude()
    freq = hilbert.instantaneous_frequency()
    ms = hilbert.marginal_spectrum()
    assert amp.shape[0] == result.imfs().n_imfs()
    assert freq.shape[0] == result.imfs().n_imfs()
    assert len(ms) > 0


# ─── Algorithm type ────────────────────────────────────────────────────


def test_algorithm_type():
    import ferromode

    signal = _make_signal()
    result = ferromode.emd(signal)
    algo = result.algorithm()
    assert algo is not None


# ─── ImfCollection ─────────────────────────────────────────────────────


def test_imf_collection_properties():
    import ferromode

    signal = _make_signal()
    result = ferromode.emd(signal)
    imfs = result.imfs()
    assert imfs.n_imfs() >= 1
    oi = imfs.orthogonality_index()
    assert oi >= 0.0
    imf_array = imfs.imfs()
    assert imf_array.shape[0] == imfs.n_imfs()
    residue = imfs.residue()
    assert residue.dtype == np.float64


# ─── Boundary condition enum ──────────────────────────────────────────


def test_boundary_condition():
    import ferromode

    bc = ferromode.BoundaryCondition("mirror_even")
    assert bc is not None


def test_boundary_condition_invalid():
    import ferromode

    with pytest.raises(ValueError):
        ferromode.BoundaryCondition("nonexistent")


# ─── Reconstruction accuracy ──────────────────────────────────────────


def test_emd_reconstruction_accuracy():
    import ferromode

    signal = _make_signal()
    result = ferromode.emd(signal)
    recon = result.reconstruct()
    max_error = np.max(np.abs(signal - recon))
    assert max_error < 1e-6, f"Reconstruction error too large: {max_error}"


def test_vmd_reconstruction_accuracy():
    import ferromode

    signal = _make_signal()
    config = ferromode.VmdConfig(n_modes=2, max_iterations=200)
    result = ferromode.vmd(signal, config)
    recon = result.reconstruct()
    max_error = np.max(np.abs(signal - recon))
    assert max_error < 1e-3, f"VMD reconstruction error too large: {max_error}"


# ─── Differentiable EMD ────────────────────────────────────────────────────


def test_emd_forward_backward():
    import ferromode

    signal = _make_signal(200)
    fwd = ferromode.emd_forward(signal, max_imfs=4)
    assert len(fwd.imfs) >= 1
    assert np.isfinite(fwd.reconstruction_error())

    grads = [[1.0] * len(signal) for _ in range(len(fwd.imfs))]
    grad = fwd.backward(grads)
    assert len(grad) == len(signal)
    assert np.all(np.isfinite(grad))


def test_emd_backward_matches_finite_difference():
    """The analytic VJP must match a finite-difference gradient of the IMF."""
    import ferromode

    signal = _make_signal(64)

    def imf0(x):
        return np.array(ferromode.emd_forward(x, max_imfs=1).imfs[0])

    # d(sum of IMF_0)/d(signal) via forward differences vs analytic backward.
    fwd = ferromode.emd_forward(signal, max_imfs=1)
    grads = [[1.0] * len(signal)]  # upstream gradient = 1 on every IMF_0 sample
    analytic = np.array(fwd.backward(grads))

    eps = 1e-5
    base = imf0(signal).sum()
    fd = np.zeros(len(signal))
    for j in range(len(signal)):
        xp = signal.copy()
        xp[j] += eps
        fd[j] = (imf0(xp).sum() - base) / eps

    assert np.max(np.abs(analytic - fd)) < 1e-4
